pkg/node/statestore_test.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package node
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/storage"
+)
+
+func TestInitStamperStore(t *testing.T) {
+	dataDir := t.TempDir()
+	stateStore, err := InitStateStore(log.Noop, dataDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ids := make(map[string]int)
+
+	// add 10 stamps to the state store
+	for i := 0; i < 10; i++ {
+		bID := make([]byte, 32)
+		_, err = rand.Read(bID)
+		if err != nil {
+			t.Fatal(err)
+		}
+		si := postage.NewStampIssuer("", "", bID, big.NewInt(3), 11, 10, 1000, true)
+		err = stateStore.Put(fmt.Sprintf("postage%s", string(si.ID())), si)
+		if err != nil {
+			t.Fatal(err)
+		}
+		ids[string(si.ID())] = 0
+	}
+
+	// init stamper store should migrate 10 stamps from state store
+	stamperStore, err := InitStamperStore(log.Noop, dataDir, stateStore)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// check stamper store has 10 stamps
+	err = stamperStore.Iterate(
+		storage.Query{
+			Factory: func() storage.Item { return new(postage.StampIssuerItem) },
+		}, func(result storage.Result) (bool, error) {
+			issuer := result.Entry.(*postage.StampIssuerItem).Issuer
+			ids[string(issuer.ID())]++
+			return false, nil
+		})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	nMissing := 0
+	for _, v := range ids {
+		if v != 1 {
+			nMissing++
+		}
+	}
+	if nMissing > 0 {
+		t.Fatalf("missing %d stamps", nMissing)
+	}
+
+	t.Cleanup(func() {
+		_ = stateStore.Close()
```

Comment:
Why are these errors ignored? Also why there is need for a sleep?

Reply:
The sleep gives levelDB some time to shut down it's goroutines.
For some reason, levelDB takes longer to close the channels and this results in gourtine leak. 

```
goleak: Errors on successful test run: found unexpected goroutines:
[Goroutine 51 in state select, with github.com/syndtr/goleveldb/leveldb.(*DB).mpoolDrain on top of the stack:
goroutine 51 [select]:
github.com/syndtr/goleveldb/leveldb.(*DB).mpoolDrain(0x14000532c40)
	/Users/macbookpro/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/db_state.go:110 +0xe4
created by github.com/syndtr/goleveldb/leveldb.openDB
	/Users/macbookpro/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/db.go:149 +0x424
```

The simpliest fix was to sleep

Reply:
So this is not the right fix actually. This might work on your machine but it may not always work on CI.

I think you will have to add some ignore calls for these leaks. Look at the storage pkg for examples.